### ISSUE TITLE
which is the right way to setup number of epochs? [NoobQuestion#2] 

### ISSUE THREAD

**Issue by IgorWojas** [mlpack/mlpack#3625]
Hello!!

Once again I come to you seeking knowledge about basic mlpack related stuff.

As the title says, Im having some trouble deciding how to properly set the number of epochs. I'm asking this mailnly because Im a typical user trying to use the product in a way it was not intended to ;)

During the training process i want to see a preview (loss is not enough to judge). I'm working with images right now so looking at a predicted image during training tells me alot about the progress. At first I was just putting the Train() function in a loop like that:

`int epochs = 1000
for (int e = 0; e < epochs; e++) {
Train(in, out, opt);
if (e % 10 == 0) {
/* show prediction*/
}
}`

When doing like so the Progress Bar prints epoch 1/1 no matter what.
Then I realised that optimizers have maxIterations parameter which gives a similar effect. So:

`ens::Adam opt(LR, batch, beta1, beta2, eps, trainInput.n_cols * 10); `

which by itself gives 10 Epochs but this is actually telling the optimizer to go through all the columns of input data 10 times. Anyway doing it like that makes the ProgessBar show the number of epochs x/10. 

When combining both methods I get a feeling that something (the optimizer?) is being reset after maxIterations is reached. The loss value drops nicely but after said "reset" it goes up. If i drew a graph of loss it would look like that: https://blackburntools.com/articles/saw-tooth-geometry/images/rip-unsloped-point.gif just sloped down a little.

Then I tried without the prediction preview at all and the training went really fast in terms of the loss decrease.

I've looked into the examples and it seems most just use the tolerance parameter to stop the training at a given point.

For now I'm just using a combination of methods mentioned above, setting the maxIterations to trainInput.n_cols * 100 so the "resets" aren't frequent and my own loop set to 10 iterations to get 1000 total epochs and some previews.

So my question is how to have the best of both worlds? I'd love to see the predicted preview every n-th epoch but I also want the fast "convergence". 

Thanks for your patience
Cheers!!!

EDIT: In this example https://github.com/mlpack/examples/blob/master/neural_network_regression/nn_regression.cpp number of epochs is set through maxIterations in the optimizer parameters. This partially answers my question.

**Comment by rcurtin** [comment]
The loss prediction that the `ProgressBar` is showing you is necessarily something of an estimate.  If we want to compute an exact loss value, we need to actually compute predictions with the current state of the neural network on *every* point in `in`---and that can be a large amount of computation!  So, we don't do this, and instead use the loss just on each minibatch of training points as the progress bar estimate (all other toolkits do something like this too).  This means that the first minibatch loss estimate has a lot of variance, and then as we add estimates for other minibatches, it slowly converges to the full-data loss (but even then it is still something of an estimate).  This explains the "jumpy" behavior you see at the beginning of the epoch.

The optimizer isn't actually being reset between iterations, and the model isn't getting worse---it's just a matter of the approximation of how we can efficiently get some kind of loss estimate.

One experiment you could do to see this is use the `PrintLoss()` callback instead of `ProgressBar()`.  If you run for 10 epochs, this will print 10 loss values.  Then, compare this with what happens if you use an adaptation of your single-epoch training loop:

```
int epochs = 1000;
for (int e = 0; e < epochs; e++)
{
  double loss = Train(in, out, opt);
  std::cout << loss << std::endl;
}
```

and this should print very similar numbers (they may not be exactly the same due to randomness in shuffling the data and other small effects).

**Comment by IgorWojas** [comment]
@rcurtin Thank you for thorough explanation!!! I have tried something similar to the code you proposed but i couldn't understand the slight differences in values. Now everything is clear. Sooo Resolved!!! 

Thanks again
Cheers!!!
